controller/controller.go

```diff
@@ -1098,6 +1098,28 @@ func (ctrl *ProvisionController) syncVolume(ctx context.Context, obj interface{}
 	if !ok {
 		return fmt.Errorf("expected volume but got %+v", obj)
 	}
+	var modifiedFinalizers []string
+	var modified bool
+	reclaimPolicy := volume.Spec.PersistentVolumeReclaimPolicy
+	// Check if the `addFinalizer` config option is disabled, i.e, rollback scenario, or the reclaim policy is changed
+	// to `Retain` or `Recycle`
+	if !ctrl.addFinalizer || reclaimPolicy == v1.PersistentVolumeReclaimRetain || reclaimPolicy == v1.PersistentVolumeReclaimRecycle {
+		modifiedFinalizers, modified = removeFinalizer(volume.ObjectMeta.Finalizers, finalizerPV)
+	}
+	// Add the finalizer only if `addFinalizer` config option is enabled, finalizer doesn't exist and PV is not already
+	// under deletion.
+	if ctrl.addFinalizer && reclaimPolicy == v1.PersistentVolumeReclaimDelete && volume.DeletionTimestamp == nil {
+		modifiedFinalizers, modified = addFinalizer(volume.ObjectMeta.Finalizers, finalizerPV)
+	}
+
+	if modified {
+		volume.ObjectMeta.Finalizers = modifiedFinalizers
+		newVolume, err := ctrl.updatePersistentVolume(ctx, volume)
+		if err != nil {
+			return fmt.Errorf("failed to modify finalizers to %+v on volume %s err: %+v", modifiedFinalizers, volume.Name, err)
+		}
+		volume = newVolume
+	}
```

Comment:
Can you please move it to a separate function like `HandleProtectionFinalizer`? I like `syncVolume` short and self-describing.

Reply:
Done.